### PR TITLE
[Merged by Bors] - Fix possible nil ptr dereference in HTTP retry check

### LINKED ISSUE #4645: Possible crash in poet client CheckRetry callback
## Description
The `checkRetry()` callback in poet client might dereference a nil pointer 

## Steps to reproduce
Unknown

## Actual Behavior
crash

## Expected Behavior
no crash, don't retry

## Environment
?

## Additional Resources 
https://github.com/spacemeshos/go-spacemesh/blob/b907b50279fc046b8fa53c7da578bb30024ef930/activation/poet.go#L49-L58

image

### PR DESCRIPTION
**Pull request by poszu**
## Motivation
Closes #4645 

## Changes
Checking for an error before dereferencing `resp`

## Test Plan
Added unit tests covering the whole `checkRetry()`.

### CODE REVIEW

File: activation/poet.go [hunk 1]
```diff
@@ -50,7 +50,9 @@ func checkRetry(ctx context.Context, resp *http.Response, err error) (bool, erro
 	if retry, err := retryablehttp.DefaultRetryPolicy(ctx, resp, err); retry || err != nil {
```
Comment (dshulyak) [inline activation/poet.go]: this method will not fail with nil resp? does it make sense to exit immediately with false is resp is nil, just for additionnal defense?

**Review by dshulyak — APPROVED** [review]

File: activation/poet.go (hunk 1 above)
Comment (fasmat) [reply activation/poet.go]: afaik `resp` is only `nil` if `err` is not, at least that's how `net/http` deals with failed requests.

That being said, we only want to overwrite the behaviour in the case of 404? Could we then write `checkRetry` like this:

```golang
func checkRetry(ctx context.Context, resp *http.Response, err error) (bool, error) {
	if resp != nil && resp.StatusCode == http.StatusNotFound {
		return true, err
	}
	return retryablehttp.DefaultRetryPolicy(ctx, resp, err)
}
```

Comment (poszu) [reply activation/poet.go]: @fasmat, much nicer, thx. Updated the code.

**Comment by poszu** [comment]
bors merge

**Comment by poszu** [comment]
`TestCheckpoint` timed out. There's nothing obvious in the logs - retrying

bors merge

**Comment by poszu** [comment]
Weird timeout in macos unittests...

bors merge

**Pull request closed without merging**